Implement several RandomVariables as SymbolicRandomVariables

ricardoV94 opened the pull request:
<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-devs/pymc/releases -->

## Description
<!--- Describe your changes in detail -->
Implement several RandomVariables as SymbolicRandomVariables

This allows sampling from multiple backends without having to dispatch for each one

Also:
* Allow signature to handle rng and size arguments explicitly.
* Move rv_op method to the SymbolicRandomVariable class and get rid of dummy inputs logic (it was needed in previous versions of PyTensor)
* Allow dispatch methods without filtering of inputs for SymbolicRandomVariable distributions


## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [ ] Closes #
- [ ] Related to #

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->


<!-- readthedocs-preview pymc start -->
----
📚 Documentation preview 📚: https://pymc--7239.org.readthedocs.build/en/7239/

<!-- readthedocs-preview pymc end -->

ricardoV94 committed:
Tweaks to SymbolicRandomVariables

* Allow signature to handle rng and size arguments explicitly.
* Parse ndim_supp and ndims_params from class signature
* Move rv_op method to the SymbolicRandomVariable class and get rid of dummy inputs logic (it was needed in previous versions of PyTensor)
* Fix errors in automatic signature of CustomDist
* Allow dispatch methods without filtering of inputs for SymbolicRandomVariable distributions

ricardoV94 committed:
Reimplement several RandomVariables as SymbolicRandomVariables

This allows sampling from multiple backends without having to dispatch for each one

ricardoV94 commented on pymc/distributions/distribution.py:
```diff
@@ -886,10 +963,10 @@ def _infer_final_signature(signature: str, n_inputs, n_updates) -> str:
             inputs, outputs = _parse_gufunc_signature(signature)
         if len(inputs) == n_inputs - 1:
```
Should also add rngs to input signature

jessegrabowski commented on pymc/distributions/continuous.py:
```diff
@@ -1236,20 +1235,33 @@ def icdf(value, alpha, beta):
 
-kumaraswamy = KumaraswamyRV()
+        if rv_size_is_none(size):
+            size = pt.as_tensor(pt.broadcast_shape(a, b))
```
This should be a helper function -- `_setup_rng(rng, size, *params)` ?

jessegrabowski commented on pymc/distributions/distribution.py:
```diff
@@ -185,30 +185,56 @@ def _random(*args, **kwargs):
+                    if isinstance(op, RandomVariable):
+                        rng, size, dtype, *dist_params = dist_params
+                    elif params_idxs:
+                        dist_params = [dist_params[i] for i in params_idxs]
```
Same comment; duplicated code

ricardoV94 replied:
I'm okay with this repetition

jessegrabowski commented on pymc/distributions/distribution.py:
```diff
@@ -267,7 +298,10 @@ class SymbolicRandomVariable(OpFromGraph):
-    """Numpy-like vectorized signature of the distribution."""
+    """Numpy-like vectorized signature of the distribution.
+
+    It allows tokens [rng], [size] to identify the special inputs.
```
Maybe some examples of what a signature should look like?

jessegrabowski commented on pymc/distributions/timeseries.py:
```diff
@@ -967,55 +998,6 @@ def dist(cls, dt, sde_fn, sde_pars, *, init_dist=None, steps=None, **kwargs):
-        # Variables with underscore suffix are dummy inputs into the OpFromGraph
-        init_ = init_dist.type()
-        sde_pars_ = [x.type() for x in sde_pars]
-        steps_ = steps.type()
```
Is this `type` thing not required anymore? I remember having to do that when I wrote distributions.

ricardoV94 replied:
That was all a side effect of not being able to use shared variables and constants as inputs. No longer needed

jessegrabowski commented on pymc/distributions/truncated.py:
```diff
@@ -71,6 +76,138 @@ def __init__(
+            return (
+                (truncated_rv, reject_draws),
+                collect_default_updates(new_truncated_rv),
+                until(~pt.any(reject_draws)),
```
Does `~` work as expected on tensor variables? I always thought you had to use `pt.bitwise_not`

ricardoV94 replied:
I think only eq and neq don't work

jessegrabowski reviewed (approved):
Seems like this is mostly shuffling around and simplifying stuff that already exists, so if tests pass it looks great. I'm a bit unclear on how the signatures for RVs work, was that changed in this PR or another one? The square brackets in particular throw me off. There's also some opportunities for refactoring duplicated code.

jessegrabowski commented on pymc/distributions/distribution.py:
```diff
@@ -279,24 +308,70 @@ class SymbolicRandomVariable(OpFromGraph):
+    def _params_signature(cls):
+        inputs_signature, outputs_signature = cls.signature.split("->")
+        inputs_params_signature = [
+            inp_sig for inp_sig in inputs_signature.split(",") if inp_sig not in ("[rng]", "[size]")
```
Maybe a global `SPECIAL_TOKENS = ["[rng]", "[size]"]` in case you want to add/remove/modify these later?

jessegrabowski replied:
I think this whole section is just to remove the special tokens before passing it to `_parse_gufunc_signature` yeah? Couldn't this be done more transparently with a regex, like `core_signature = re.sub(special_token_pattern, '', cls.signature)`?

ricardoV94 replied:
Cross that bridge when/if we get there?

jessegrabowski replied:
sure

ricardoV94 replied:
> I think this whole section is just to remove the special tokens before passing it to _parse_gufunc_signature yeah? Couldn't this be done more transparently with a regex, like core_signature = re.sub(special_token_pattern, '', cls.signature)?

Needs to be able to get rid of the optional/useless commas, but otherwise I'm not against regex, just not my zone.

ricardoV94 replied:
Done the regex way

ricardoV94 commented on pymc/distributions/distribution.py:
```diff
@@ -244,6 +275,21 @@ def fn(*args, **kwargs):
+        descr_get = super().__get__ if instance is None else self.__func__.__get__
+        return descr_get(instance, type_)
+
+
```
Some bit of blackmagic to allow dynamic properties from the signature, to work both when signatures are class attributes or only instance attributes. Let me know if you see a better/ less magical solution.

jessegrabowski replied:
Why is it needed though?

ricardoV94 replied:
Sometimes we have the signature fixed in the class (DM), sometimes only known at runtime (Mixture). This allows us to parse ndims_params, ndim_supp, default_outputs from the signature regardless of when it's defined.

jessegrabowski replied:
Would it be better to have a class_signature (always defined) and a instance_signature (default None), and drop back to the class_signature if instance_signature is None?

ricardoV94 replied:
A class signature cannot always be defined. The point is I want to extract `ndim_supp` from either a class signature (without having to initiate said class) or, if it cannot be known in advance, from an instance signature. I need a static class method or instance method that can read either the class or instance signature, this achieves exactly that.

ricardoV94 replied:
Do I really need this? Not so much. Alternatives are:
1. Hard code ndim_supp/ndims_params/default_output when they can be known in advance. It's just redundant/error prone because all info exists in the signature.
2. Don't provide any of these until the Op is defined. Drawback? I can't use these lines as straightfoward:

We have to create dummy Ops with empty size just just to find the ndim_supp:
https://github.com/pymc-devs/pymc/blob/5f95fc292adebbeb31b6de825b7b407042e382fd/pymc/distributions/distribution.py#L486-L489

I can't write classmethods of the new SymbolicRandomVariables like this:
```python
        if rv_size_is_none(size):
            size = implicit_size_from_params(b, kappa, mu, ndims_params=cls.ndims_params)
```
Because cls.ndims_params doesn't exist until we create the `Op` below. 

I could obviously parse ndims_params/ndim_supp on the spot from the signature. It's just a small convenience. But OTOH it's also just 5 lines of Python

ricardoV94 replied:
@jessegrabowski My last commit reverts the magic solution for the cleanest alternative I can think of. Which one do you prefer?

ricardoV94 replied:
I removed the non-magic commit, for reference it's here: https://github.com/pymc-devs/pymc/commit/20e6e3ba23b27020ce99a603394d8b67b8fef64a

ricardoV94 committed:
Remove wrong instantiation from test

This is not only not used anywhere, but could also try to initiate an RV with incompatible shapes and size
